### PR TITLE
Workarounds for "common" (aka axis) hover label clipping

### PR DESCRIPTION
**Pull request by etpinard**
resolves https://github.com/plotly/plotly.js/issues/4221

---

https://github.com/plotly/plotly.js/commit/68521df0bfb6062c9159bbdfd6a2b269b6c04c9b - Under `hovermode: 'x'`, the common hover label (i.e. the one on top of the x axis) is constrained into the viewport while the hover label arrow appear on the left or right edge.

Peek 2019-10-22 14-48

- https://codepen.io/etpinard/pen/qBBrLvp
- https://codepen.io/etpinard/pen/RwwpPLb (with `xaxis.automargin: true`)

----

https://github.com/plotly/plotly.js/commit/2c94e5f5845127b13cbae6034168359f740d2b49 - Under `hovermode: 'y'`, the common hover label for `side:'left'` y axes has its text shifted to the right so that its beginning is visible and then clipped so that it doesn't overlap the hover label arrow.

Peek 2019-10-22 14-51

- https://codepen.io/etpinard/pen/eYYvbog
- https://codepen.io/etpinard/pen/qBBrLwx (with `yaxis.automargin: true`)

---

cc @plotly/plotly_js

### CODE REVIEW

File: src/components/fx/hover.js [hunk 1]
```diff
@@ -790,41 +790,103 @@ function createHoverText(hoverData, opts, gd) {
...
 
-            label.attr('transform', 'translate(' +
-                (xa._offset + (ya.side === 'right' ? xa._length : 0)) + ',' +
-                (ya._offset + (c0.y0 + c0.y1) / 2) + ')');
+            var halfHeight = tbb.height / 2;
+            var clipId = 'clip' + fullLayout._uid + 'commonlabel' + ya._id;
+            var clipPath;
+            var ltx;
+
+            if(lx < (tbb.width + 2 * HOVERTEXTPAD + HOVERARROWSIZE)) {
```
Comment (etpinard) [inline src/components/fx/hover.js]: @nicolaskruchten this use to be 

```js
if(lx < (tbb.width + HOVERTEXTPAD + HOVERARROWSIZE)) { // ...
```

which may have caused the problem described in https://github.com/plotly/plotly.js/issues/4221#issuecomment-545163444

New attempt: https://codepen.io/etpinard/pen/qBBrLwx

**Comment by archmoj** [comment]
Very nice.
The only issue I noticed is one has multi-line labels the short lines may disappear like this [codepen](https://codepen.io/MojtabaSamimi/pen/BaaWbLR?editors=1000).

**Comment by etpinard** [comment]
> The only issue I noticed is one has multi-line labels the short lines may disappear like this [codepen](https://codepen.io/MojtabaSamimi/pen/BaaWbLR?editors=1000).

@archmoj Thanks very much for spotting that! Here's how it looks like after the latest commit:

Peek 2019-10-24 15-09

https://codepen.io/etpinard/pen/eYYWyOw?editors=1010

File: src/components/fx/hover.js [hunk 2]
```diff
@@ -772,18 +772,19 @@ function createHoverText(hoverData, opts, gd) {
         var commonBgColor = commonLabelOpts.bgcolor || Color.defaultLine;
         var commonStroke = commonLabelOpts.bordercolor || Color.contrast(commonBgColor);
         var contrastColor = Color.contrast(commonBgColor);
+        var commonLabelFont = {
+            family: commonLabelOpts.font.family || fontFamily,
+            size: commonLabelOpts.font.size || fontSize,
+            color: commonLabelOpts.font.color || contrastColor
```
Comment (archmoj) [inline src/components/fx/hover.js]: Do we always have `commonLabelOpts.font` in the options?
If not, we may use 
```js
(commonLabelOpts.font || {}).etc
```.

Comment (etpinard) [reply src/components/fx/hover.js]: Yep! If not many things would have been broken for a long time.

**Comment by archmoj** [comment]
Nicely done.
:dancer: :dancer: :dancer:

**Comment by etpinard** [comment]
Thanks again for the review @archmoj - merging!

**Pull request merged**